tests/erc721/erc721_test.go

```diff
@@ -162,6 +171,289 @@ func TestOpenSeaProxyApproval(t *testing.T) {
+				got, err := nft.IsApprovedForAll(nil, toAddr, sim.Addr(proxy))
+				if err != nil || got != tt.wantPreApproved {
+					t.Errorf("IsApprovedForAll([owner], [proxy]) got %t, err = %v; want %t, nil err", got, err, tt.wantPreApproved)
+				}
+			}
+		})
+	}
+}
+
+func TestOpenSeaProxyPreApprovalOptInOut(t *testing.T) {
+	sim, _, _ := deploy(t)
+
+	tests := []struct {
+		name         string
+		mints        bool
+		hasProxy     bool
+		wantApproved bool
+		wantEvents   []*ERC721ApprovalForAll
+	}{
+		{
+			name:         "opt in with proxy after mint",
+			mints:        true,
+			hasProxy:     true,
+			wantApproved: true,
+			wantEvents: []*ERC721ApprovalForAll{
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: true,
+				},
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: true,
+				},
+			},
+		},
+		{
+			name:         "opt out with proxy after mint",
+			mints:        true,
+			hasProxy:     true,
+			wantApproved: false,
+			wantEvents: []*ERC721ApprovalForAll{
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: true,
+				},
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: false,
+				},
+			},
+		},
+		{
+			name:         "opt in with proxy",
+			mints:        false,
+			hasProxy:     true,
+			wantApproved: true,
+			wantEvents: []*ERC721ApprovalForAll{
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: true,
+				},
+			},
+		},
+		{
+			name:         "opt out with proxy",
+			mints:        false,
+			hasProxy:     true,
+			wantApproved: false,
+			wantEvents: []*ERC721ApprovalForAll{
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: false,
+				},
+			},
+		},
+		{
+			name:         "opt in after mint",
+			mints:        true,
+			hasProxy:     false,
+			wantApproved: true,
+			wantEvents: []*ERC721ApprovalForAll{
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: true,
+				},
+			},
+		},
+		{
+			name:         "opt out after mint",
+			mints:        true,
+			hasProxy:     false,
+			wantApproved: false,
+			wantEvents: []*ERC721ApprovalForAll{
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: false,
+				},
+			},
+		},
+		{
+			name:         "opt in",
+			mints:        false,
+			hasProxy:     false,
+			wantApproved: true,
+			wantEvents: []*ERC721ApprovalForAll{
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: true,
+				},
+			},
+		},
+		{
+			name:         "opt out",
+			mints:        false,
+			hasProxy:     false,
+			wantApproved: false,
+			wantEvents: []*ERC721ApprovalForAll{
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: false,
+				},
+			},
+		},
+	}
```

Comment:
I think we have a case that's impossible to catch, in which someone doesn't explicitly opt in but then later sets a proxy. Nothing we can really do about that, yeah?

Reply:
No we can absolute catch this by adding something like `hasProxyAfterMint` to `TestOpenSeaProxyPreApprovalOptInOut`

Reply:
oh I think I misunderstood earlier. I added an extra `doSetApproved` flag to the above that determines if the manual opt in is actually called. This way we can have something like
```go
{
	name:                 "creating proxy after minting without proxy",
	hasProxyBeforeMint:   false,
	mints:                true,
	createProxyAfterMint: true,
	doSetApproved:        false,
	wantApproved:         false,
	wantEvents:           nil,
},
```
event thought the table gets quite long now.